### PR TITLE
Add support for IExecutable to Data

### PR DESCRIPTION
**Pull request by PascalSenn**

### CODE REVIEW

File: src/HotChocolate/Data/src/Data/Extensions/DataEnumerableExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+using System.Collections.Generic;
+using System.Linq;
+
+namespace HotChocolate.Data
+{
+    public static class DataEnumerableExtensions
+    {
+        public static QueryableExecutable<T> AsExecutable<T>(this IEnumerable<T> source)
```
Comment (michaelstaib) [inline src/HotChocolate/Data/src/Data/Extensions/DataEnumerableExtensions.cs:8]: documentation

File: src/HotChocolate/Core/src/Abstractions/IExecutable.cs [hunk 2]
```diff
@@ -1,12 +1,53 @@
...
+    /// Represents a abstract executable that is well known in the framework. If the execution
+    /// engine encounters a <see cref="IExecutable"/>, it will call execute it
+    /// </summary>
     public interface IExecutable
     {
-        ValueTask<object> ExecuteAsync(CancellationToken cancellationToken);
+        /// <summary>
+        /// The current state of the executable
+        /// </summary>
+        object Source { get; }
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Abstractions/IExecutable.cs:18]: Do we need an abstract source? or is it enough to have it on the specific implementations?

File: src/HotChocolate/Data/src/Data/IQueryableExecutable.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
+using System.Linq;
+
+namespace HotChocolate.Data
+{
+    public interface IQueryableExecutable<T> : IExecutable<T>
```
Comment (michaelstaib) [inline src/HotChocolate/Data/src/Data/IQueryableExecutable.cs:5]: Do we need an interface here or is QueryableExecutable enough?

File: src/HotChocolate/Data/src/Data/QueryableExecutable.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        public bool InMemory { get; }
+
+        public QueryableExecutable<T> WithSource(IQueryable<T> source)
+        {
+            return new QueryableExecutable<T>(source);
+        }
+
+        public virtual ValueTask<IList> ToListAsync(CancellationToken cancellationToken)
+        {
+            return new ValueTask<IList>(Source.ToList());
```
Comment (michaelstaib) [inline src/HotChocolate/Data/src/Data/QueryableExecutable.cs:29]: This cannot be the correct implementation or?

Comment (michaelstaib) [reply src/HotChocolate/Data/src/Data/QueryableExecutable.cs:29]: Even if we have a ef specific implementation there are other IQueryable drivers where we need to execute async like mongo, or ravendb and so on. So the queryable executable must check if it is in-mem and only then use sync in all other cases we must check if IAsyncEnumerable is possible or if we need to spin up a task.

File: src/HotChocolate/Data/src/Data/IQueryableExecutable.cs (hunk 3 above)
Comment (PascalSenn) [reply src/HotChocolate/Data/src/Data/IQueryableExecutable.cs:5]: Should i remove it?

File: src/HotChocolate/Core/src/Abstractions/IExecutable.cs (hunk 2 above)
Comment (PascalSenn) [reply src/HotChocolate/Core/src/Abstractions/IExecutable.cs:18]: I guess it makes sense to have it there. We use it in Cursor Paging and Offset Paging

**Comment by PascalSenn** [comment]
@michaelstaib  i also fixed the behavior of SingleOrDefault and FirstOrDefault. They can both return null when the collections is empy

**Review by michaelstaib — APPROVED** [review]

**Review by michaelstaib — APPROVED** [review]

**Review by michaelstaib — APPROVED** [review]

**Pull request merged**